apollo-federation/src/sources/connect/validation/schema.rs

```diff
@@ -327,14 +345,145 @@ fn advanced_validations(schema: &SchemaInfo, subgraph_name: &str) -> Vec<Message
     entity_checker.check_for_missing_entity_connectors(schema)
 }
 
+fn compute_input_trie(connector: &Connector) -> SelectionTrie {
+    let mut merged = SelectionTrie::new();
+    for param in connector.variable_references() {
+        merged.extend(&param.selection);
+    }
+    merged
+}
+
+struct SelectionSetWalker<'walker> {
+    trie: &'walker SelectionTrie,
+    unmapped_fields: HashSet<String>,
+}
+
+impl<'walker> SelectionSetWalker<'walker> {
+    fn new(trie: &'walker SelectionTrie) -> Self {
+        SelectionSetWalker {
+            trie,
+            unmapped_fields: HashSet::new(),
+        }
+    }
+}
+
+#[derive(Debug, thiserror::Error)]
+enum ShapeVisitorError<'error> {
+    #[error(
+        "The `@connect` directive on `{connector}` specifies a batch entity resolver, but the key `{unset}` is not a subset of the output shape `{output_shape}`"
+    )]
+    BatchKeyNotSubsetOfOutputShape {
+        connector: String,
+        unset: &'error String,
+        output_shape: &'error Shape,
+    },
+    #[error("Attempted to resolve key on unexpected shape `{shape_str}`")]
+    UnexpectedKeyOnShape { shape_str: String },
+    #[error("Batch keys cannot be mapped from non-`$root` sources (i.e. `$context`, `$this`)")]
```

Comment:
maybe something about how the fields matching the `$batch` variable must come from the API response? (again not exposing the term "$root" to users)